template/web/src/actions/get-collection-metadata.ts

```diff
@@ -0,0 +1,125 @@
+// Minimal ABI with contractURI view function
+import { Abi, Address } from 'abitype';
+import { PublicClient } from 'viem';
+import { ipfsToHTTP } from '@/utils/ipfs';
+
+const AbiFunctionContractURI = [
+  {
+    inputs: [],
+    name: 'contractURI',
+    outputs: [{ internalType: 'string', name: '', type: 'string' }],
+    stateMutability: 'view',
+    type: 'function',
+  },
+] as const;
+
+// Minimal ABI with uri view function
+const AbiFunctionWithUri = [
+  {
+    inputs: [{ internalType: 'uint256', name: 'id', type: 'uint256' }],
+    name: 'uri',
+    outputs: [{ internalType: 'string', name: '', type: 'string' }],
+    stateMutability: 'view',
+    type: 'function',
+  },
+] as const;
+
+/**
+ * There is some differences in URI standards between ERC721 and 1155, handle those in this component.
+ */
+const uriFunctionType = {
+  uri: 'uri',
+  contractURI: 'contractURI',
+} as const;
+
+const getIsAbiWithContractURIViewFunction = (abi: Abi) =>
+  !!abi.find(
+    (el) => el.type === 'function' && el.name === uriFunctionType.contractURI,
+  ) as unknown as typeof AbiFunctionContractURI;
+const getIsAbiWithUriViewFunction = (abi: Abi) =>
+  !!abi.find(
+    (el) => el.type === 'function' && el.name === uriFunctionType.uri,
+  ) as unknown as typeof AbiFunctionWithUri;
+
+const isValidAbi = (abi: Abi) =>
+  getIsAbiWithContractURIViewFunction(abi) || getIsAbiWithUriViewFunction(abi);
+
+export type JsonMetadata = {
```

Comment:
This file has a lot of things going on. Also, I wonder if a lot of this code should just live in OnchainKit.

@Yuripetusko what do you think?

Reply:
Yes @snehkoul-cb noted the same, and rightly so. However when I tried to decouple this I found that each util depends on each other too much. It mostly comes due to the fact that this is trying to get metdata uri from 2 separate functions with different signatures. And typescript doesn't have a way to create type guards for arrays like ABI, so we can't have a union type that it would understand when passing to viem/wagmi. So there's a lot of weird code because of this.

I wonder if this should just be split into separate functions and users would have to use one or the other depending on type of NFT contract (one with `uri` and another with `contractURI` 🤔)

Re: moving this to onchainkit. Yes I think absolutly + possibly the ipfs utils changes I did in another PR here?

Reply:
Yeah the more I think about this the more it makes sense. Most of the code in this file is trying to guess what kind of ABI it is, and then depending on the type, passing different params to viem readContract function. And it's not very easy to move this code into utils because of typescript limitations. So I wonder if instead of trying to make a universal function that checks what kind of abi it is, we just create 2 hooks `useContractUriMetadata` and `useUriMetadata` or something along the lines

Reply:
or `useErc1155Metadata` (with optional tokenId argument) and `useNftCollectionMetadata` (purely fro `contractURI`)

Reply:
ok I think this is the way to go actually, I'll try to make these changes

Reply:
Ok @Zizzamia hopefully it's more readable now. But I should also probably move this to OnChainkit soon